### PR TITLE
[DistillationTrainer refactor] Heavy-clean the Liger path: share extraction with the chunked path

### PR DESCRIPTION
**Pull request by qgallouedec**
*Item 34 (Group E) — stacked on PR 33. Part of #6449.*

Keeps the Liger fused-JSD fast path (`use_liger_kernel=True`) but heavy-cleans it: the two loss paths now **share the same hidden-state extraction and differ only in the final loss call**. Net −74 lines.

`compute_loss` always redirects `_compute_loss` through `_forward_redirection` (from item 33). `_compute_loss` extracts student hidden states (inside the student wrapper) + teacher hidden states (via the teacher's own redirection), then:
- `if self.use_liger_loss:` → `self.liger_loss(...)` over the flattened hidden states, with `true_labels` built from `completion_mask` (hard-loss weight is 0, so only the ignore-mask matters; no shift — `_get_last_hidden_state` is already completion-aligned), wrapped in the shared `maybe_gather_lm_head_ctx` for ZeRO-3, then the `num_items` global rescale.
- else → `_chunked_divergence_loss(...)`.

Deleted (all superseded by the shared extraction + shared gather util):
- `_liger_student_forward`, `_compute_liger_loss`, `_get_liger_zero3_lm_head_gather_ctx` (and its wrapping in `training_step`/`prediction_step`);
- the `hasattr`/`getattr` decoder/bias probing;
- the unused `nullcontext` import.

No `use_liger_kernel` raise (reversing the earlier "delete Liger" plan — SFT parity, and the path is verified working). The config field stays.

**Test:** `test_distillation_trainer_with_liger` (train-smoke) replaced with `test_liger_loss_agrees_with_chunked` — asserts the Liger loss equals the chunked loss on a fixed batch (`rtol=1e-3`), toggling `use_liger_loss` on one un-patched model so only the loss kernel differs. Suite green (52).

Known limitation (pre-existing, unchanged): the Liger kernel does not apply `logit_scale`/`final_logit_softcapping`, so on Gemma/Cohere it diverges slightly from the chunked path; the chunked default is exact. Worth a follow-up if Liger + those families matter.

---

> [!NOTE]
> **Medium Risk**
> Changes core distillation loss wiring for both Liger and the unified `_forward_redirection` path; behavior should be equivalent but distributed training (FSDP2/ZeRO-3) and gradient-accum normalization are touched.
> 
> **Overview**
> **Refactors `DistillationTrainer` so the Liger fused-JSD path and the default chunked path share one hidden-state extraction pipeline** and differ only in the final loss call (~74 lines removed).
> 
> `compute_loss` always routes `_compute_loss` through `_forward_redirection` (no early Liger branch). Inside `_compute_loss`, student/teacher hidden states are extracted once; when `use_liger_loss` is set, `LigerFusedLinearJSDLoss` runs on those tensors with `true_labels` from `completion_mask`, FSDP2 `full_tensor()` on lm heads, shared `maybe_gather_lm_head_ctx` for ZeRO-3, and the same `num_items_in_batch` rescale—otherwise `_chunked_divergence_loss` runs as before.
> 
> Removed: `_liger_student_forward`, `_compute_liger_loss`, `_get_liger_zero3_lm_head_gather_ctx`, and ZeRO-3 gather wrappers in `training_step` / `prediction_step`.
> 
> **Tests:** `test_distillation_trainer_with_liger` (train smoke) is replaced by `test_liger_loss_agrees_with_chunked`, which asserts Liger and chunked losses match on a fixed batch (`rtol=1e-3`) on the same model.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 8e3547fe8c019f8878c5e73210f8bf38aa9f17ef. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/experimental/distillation/distillation_trainer.py [hunk 1]
```diff
@@ -1062,11 +1061,7 @@ def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=N
         if self.model.training and inputs.get("num_items_in_batch") is not None:
             num_items_in_batch = inputs["num_items_in_batch"].clamp(min=1.0) / self.accelerator.num_processes
 
-        if self.use_liger_loss:
-            loss = self._compute_liger_loss(model, inputs, num_items_in_batch=num_items_in_batch)
-            return (loss, None) if return_outputs else loss
-
-        # Route the whole chunked loss (backbone + `lm_head` projection + JSD) through the DDP/FSDP wrapper via
+        # Route the whole loss (backbone + `lm_head` projection + JSD) through the DDP/FSDP wrapper via
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/distillation/distillation_trainer.py:1078]: Leaving as-is: distillation requires a shared vocab, which in practice means a same-architecture teacher, so the student- and teacher-forward signatures match and the student-derived check is correct. A cross-architecture, same-vocab teacher is an unsupported edge case (noted in Known-limitations); not worth the extra per-model bookkeeping.
